### ISSUE TITLE
Continuous WebJob handling ServiceBus messages fails repeatedly 

### ISSUE THREAD

**Issue by mateas** [Azure/azure-webjobs-sdk#953]
#### Problem description
We are building an architecture with WebJobs subscribing on a number of Service bus topics using `ServiceBusTrigger` attribute. During increasing load the message processing slows down and finally fails when processing one message at a time. The WebJobs crashes with System.TimeoutException accessing Azure storage. Here is a part of the job log showing the job starts up, register bindings of the topic subscribers, starts processing ServiceBus messages, throws exception and then restarts. The pattern repeats and the web app and job keeps the CPU 100% busy until the subscription messages are deleted. 

#### Job Log:
```
...
[12/23/2016 04:50:46 > 7409d1: ERR ] --- End of stack trace from previous location where exception was thrown ---
[12/23/2016 04:50:46 > 7409d1: ERR ]    at Microsoft.Azure.WebJobs.Host.Timers.BackgroundExceptionDispatcher.<>c__DisplayClass1.<Throw>b__0()
[12/23/2016 04:50:46 > 7409d1: ERR ]    at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
[12/23/2016 04:50:46 > 7409d1: ERR ]    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12/23/2016 04:50:46 > 7409d1: ERR ]    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12/23/2016 04:50:46 > 7409d1: ERR ]    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
[12/23/2016 04:50:46 > 7409d1: ERR ]    at System.Threading.ThreadHelper.ThreadStart()
[12/23/2016 04:50:46 > 7409d1: SYS ERR ] Job failed due to exit code -532462766
[12/23/2016 04:50:46 > 7409d1: SYS INFO] Process went down, waiting for 0 seconds
[12/23/2016 04:50:46 > 7409d1: SYS INFO] Status changed to PendingRestart
[12/23/2016 04:50:47 > 7409d1: SYS INFO] Run script 'PaymentModule.EventHandlerJob.exe' with script host - 'WindowsScriptHost'
[12/23/2016 04:50:47 > 7409d1: SYS INFO] Status changed to Running
[12/23/2016 04:50:53 > 7409d1: INFO] Found the following functions:
[12/23/2016 04:50:53 > 7409d1: INFO] PaymentModule.EventHandlerJob.Listeners.ArticleTopicListener.Listen
[12/23/2016 04:50:53 > 7409d1: INFO] PaymentModule.EventHandlerJob.Listeners.ArticleTopicListener.HandleModuleTestEvents
[12/23/2016 04:50:53 > 7409d1: INFO] PaymentModule.EventHandlerJob.Listeners.CustomerTopicListener.Listen
[12/23/2016 04:50:53 > 7409d1: INFO] PaymentModule.EventHandlerJob.Listeners.BookingTopicListener.Listen
[12/23/2016 04:50:53 > 7409d1: INFO] PaymentModule.EventHandlerJob.Listeners.PaymentTopicListener.Listen
[12/23/2016 04:50:53 > 7409d1: INFO] PaymentModule.EventHandlerJob.Listeners.StoreTopicListener.Listen
[12/23/2016 04:50:57 > 7409d1: INFO] Executing: 'CustomerTopicListener.Listen' - Reason: 'New ServiceBus message detected on 'CustomerModule/Subscriptions/PaymentModule.EventHandlerJob'.'
[12/23/2016 04:51:00 > 7409d1: INFO] Executing: 'BookingTopicListener.Listen' - Reason: 'New ServiceBus message detected on 'BookingModule/Subscriptions/PaymentModule.EventHandlerJob'.'
[12/23/2016 04:51:07 > 7409d1: INFO] Executing: 'PaymentTopicListener.Listen' - Reason: 'New ServiceBus message detected on 'PaymentModule/Subscriptions/PaymentModule.EventHandlerJob'.'
[12/23/2016 04:51:07 > 7409d1: INFO] Job host started
[12/23/2016 04:51:07 > 7409d1: INFO] Executed: 'CustomerTopicListener.Listen' (Succeeded)
[12/23/2016 04:51:07 > 7409d1: INFO] Executing: 'StoreTopicListener.Listen' - Reason: 'New ServiceBus message detected on 'FacilityModule/Subscriptions/PaymentModule.EventHandlerJob'.'
[12/23/2016 05:12:51 > 7409d1: ERR ] 
[12/23/2016 05:12:51 > 7409d1: ERR ] Unhandled Exception: Microsoft.WindowsAzure.Storage.StorageException: The client could not finish the operation within specified timeout. ---> System.TimeoutException: The client could not finish the operation within specified timeout.
[12/23/2016 05:12:51 > 7409d1: ERR ]    --- End of inner exception stack trace ---
[12/23/2016 05:12:51 > 7409d1: ERR ]    at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.EndExecuteAsync[T](IAsyncResult result) in c:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\ClassLibraryCommon\Core\Executor\Executor.cs:line 50
[12/23/2016 05:12:51 > 7409d1: ERR ]    at Microsoft.WindowsAzure.Storage.Queue.CloudQueue.EndExists(IAsyncResult asyncResult) in c:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\ClassLibraryCommon\Queue\CloudQueue.cs:line 1027
[12/23/2016 05:12:51 > 7409d1: ERR ]    at Microsoft.WindowsAzure.Storage.Core.Util.AsyncExtensions.<>c__DisplayClass1`1.<CreateCallback>b__0(IAsyncResult ar) in c:\Program Files (x86)\Jenkins\workspace\release_dotnet_master\Lib\ClassLibraryCommon\Core\Util\AsyncExtensions.cs:line 66
[12/23/2016 05:12:51 > 7409d1: ERR ] --- End of stack trace from previous location where exception was thrown ---
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at Microsoft.Azure.WebJobs.Host.Queues.Listeners.QueueListener.<ExecuteAsync>d__4.MoveNext()
[12/23/2016 05:12:51 > 7409d1: ERR ] --- End of stack trace from previous location where exception was thrown ---
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at Microsoft.Azure.WebJobs.Host.Timers.TaskSeriesTimer.<RunAsync>d__d.MoveNext()
[12/23/2016 05:12:51 > 7409d1: ERR ] --- End of stack trace from previous location where exception was thrown ---
[12/23/2016 05:12:51 > 7409d1: ERR ]    at Microsoft.Azure.WebJobs.Host.Timers.BackgroundExceptionDispatcher.<>c__DisplayClass1.<Throw>b__0()
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
[12/23/2016 05:12:51 > 7409d1: ERR ]    at System.Threading.ThreadHelper.ThreadStart()
[12/23/2016 05:12:51 > 7409d1: SYS ERR ] Job failed due to exit code -532462766
[12/23/2016 05:12:51 > 7409d1: SYS INFO] Process went down, waiting for 0 seconds
[12/23/2016 05:12:51 > 7409d1: SYS INFO] Status changed to PendingRestart
[12/23/2016 05:12:51 > 7409d1: SYS INFO] Run script 'PaymentModule.EventHandlerJob.exe' with script host - 'WindowsScriptHost'
[12/23/2016 05:12:51 > 7409d1: SYS INFO] Status changed to Running
...
```

#### Related information 

We are running versions:

- `<package id="Microsoft.Azure.WebJobs" version="1.1.2" targetFramework="net452" />`
- `<package id="Microsoft.Azure.WebJobs.Core" version="1.1.2" targetFramework="net452" />`
- `<package id="Microsoft.Azure.WebJobs.ServiceBus" version="1.1.2" targetFramework="net452" />`
- `<package id="WindowsAzure.ServiceBus" version="3.4.3" targetFramework="net452" />`
- `<package id="WindowsAzure.Storage" version="7.2.1" targetFramework="net452" />`

Any clues what we are doing wrong?

**Comment by mateas** [comment]
We are still experience this from time to time. Has anyone any suggestion what can be wrong?

**Comment by brettsam** [comment]
This looks like an issue that you can run into when the functions are pegging the CPU -- threads start struggling to get CPU time and background calls like this can fail.

A few questions to see if we can narrow it down:
1. Are any of your functions using a Queue trigger? That appears to be what is failing here.
2. It looks like you have 3 functions executing when this fails (at least in the example above). Are all of these using async/await patterns?
3. How much work is each of your functions doing? Do you expect them to return quickly, or are you expecting long-running, high-CPU executions from a single message?

**Comment by Zoidborg7** [comment]
I have a similar error, but I do not believe it is related to CPU load. I have a few web jobs that consume event hub messages in batches. At times, there are delays to messages being consumed, and I find these TimeoutException errors in the diagnostic log. Several every hour.

After some analysis over the past week, I believe it is caused by the mechanisms that upload webjob dashboard data. I disabled all dashboard interaction by setting the JobHostConfiguration.DashboardConnectionString to null. Since I did this for all my webjobs, I have had no further time delay gaps in the processing of the event hub messages.

I am using the 2.00-beta2 versions of the WebJob SDK and Service Bus package. All non-webjob-SDK nuget packages are the latest production release as of 1/18/2017.

The attached log file excerpt shows the exception detail for when the TimeoutException occurs.

[WebJobException.csv.txt](https://github.com/Azure/azure-webjobs-sdk/files/717419/WebJobException.csv.txt)
